ergo-lib/src/wallet/signing.rs

```diff
@@ -49,11 +49,23 @@ pub struct TransactionContext {
     data_boxes: Option<TxIoVec<ErgoBox>>,
 }
 
-impl TransactionContext {
+/// Ergo Transaction that made TransactionContext from Transaction and UnsignedTransaction
+pub trait ErgoTransaction<T> {
```

Comment:
It does not make much sense for the transaction to construct its context from itself and additional pieces of information (boxes) that it's not responsible for.

I was talking about something like this:
```rust

pub struct TransactionContext<T: ErgoTransaction> {
    /// Transaction (signed or unsigned) that spending `boxes_to_spend` (e.g has them in it's inputs) and has `data_boxes` in it's data inputs
    pub spending_tx: T,
    /// Boxes corresponding to tx inputs
    boxes_to_spend: TxIoVec<ErgoBox>,
    /// Boxes corresponding to tx data inputs
    data_boxes: Option<TxIoVec<ErgoBox>>,
}

impl<T: ErgoTransaction> TransactionContext<T> {

    fn new(
        spending_tx: T,
        boxes_to_spend: TxIoVec<ErgoBox>,
        data_boxes: Option<TxIoVec<ErgoBox>>,
        ) -> Result<Self, TxSigningError> {
        /// impl against generic tx (ErgoTransaction trait)
    }

}

pub trait ErgoTransaction {
    pub fn input_ids(&self) -> TxIoVec<BoxId>;
    pub fn data_inputs(&self) -> Option<TxIoVec<DataInput>>;
}

impl ErgoTransaction for UnsignedTransaction {
    ...
}

impl ErgoTransaction for Transaction {
    ...
}
```

Reply:
That's right, I fix it in the new commit.